### ISSUE TITLE
[CS2] Array destructuring bug

### ISSUE THREAD

**Issue by boris-petrov** [jashkenas/coffeescript#4657]
```coffeescript
x = ([a..., b]) ->
  console.log a, b

x [1, 2, 3, 4]
```

This produces the incorrect:

```javascript
var x,
  slice = [].slice;

x = function(arg) {
  var arg, i;
  a = 2 <= arg.length ? slice.call(arg, 0, i = arg.length - 1) : (i = 0, []), b = arg[i++];
  return console.log(a, b);
};

x([1, 2, 3, 4]);
```

As you can see, the variable `a` is not defined anywhere and this blows up.

Using CoffeeScript 2.0.0-beta4.

PS. Actually `b` is not defined either.

**Comment by vendethiel** [comment]
Fixed by #4640. Not yet released.

**Comment by boris-petrov** [comment]
@vendethiel - is this supposed to be fixed in beta5? I have a case where it doesn't work with it. Should I try to create a small reproduction or the fix for this is not merged in beta5?

**Comment by vendethiel** [comment]
Please try to create a repro

**Comment by helixbass** [comment]
@GeoffreyBooth this seems to have been introduced by #4478. Should be reopened

`([a..., b]) ->` illustrates the problem - `a` and `b` aren't declared in the compilation:
```
var slice = [].slice;

(function(arg) {
  var arg, i;
  a = 2 <= arg.length ? slice.call(arg, 0, i = arg.length - 1) : (i = 0, []), b = arg[i++];
});
```

**Comment by helixbass** [comment]
Should be fixed by #4683

**Comment by boris-petrov** [comment]
Thanks for the quick fixes. I'll wait for beta6 then, hopefully it comes out soon. :)

**Comment by GeoffreyBooth** [comment]
Fixed by #4683.
